### PR TITLE
Allow sim_fibo_ad to run without wells.

### PR DESCRIPTION
**Pull request by dr-robertk**
This PR fixes all places where the code was not acknowledging that the wells pointer could be zero when no wells were specified in the deck. The results for SPE9 are identical. From that I deduct that also Norne still runs ;-). I would appreciate  if somebody could take a look at this.

### CODE REVIEW

**Comment by blattms** [comment]
Jus saw that you changed from holding a reference to the wells to using a pointer. It seems like one could query the number of wells (wells_->number_of_wells)  from Wells. Therefore I would have assumed that one could use it to represent 0 wells and stick to the reference. What am I missing?

**Comment by dr-robertk** [comment]
When there is no wells in the deck then the wells_ pointer is zero. So you can't access number_of_wells. Otherwise, I would have preferred that option too.

**Comment by atgeirr** [comment]
We could however change that (zero pointer when no wells) in the WellsManager fairly easily, couldn't we? Then we could use references throughout the code. Most functions that take a pointer did so originally because wells were optional: you could drive a run with boundary conditions instead of wells. However, for this simulator we only support wells, and the explicit possibility to pass a zero pointer is not really improving readability (we must then check the number of wells in a few places instead of course).

**Comment by bska** [comment]
> for this simulator we only support wells

To be honest, I'd prefer to keep it that way.  At least until we have something resembling boundary conditions.  I suppose you could count gravity as a driving force, but outside very academic settings (e.g., the venerable rising gas bubble test case) that is of limited utility.

**Comment by dr-robertk** [comment]
@bska: The reason for that feature is to have a simple test without wells for testing the simulator in parallel. The reduction of complexity is necessary in my point of view.

**Comment by dr-robertk** [comment]
@atgeirr: If we change the WellsManager to export references, then I would proposed to replace the WellsManager in opm-autodiff with the implementation from opm-parser. This will also reduce maintenance, one implementation instead of two. One the other hand, I'm not sure if this is a good idea right now. But when is a good time? Probably never.

File: opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp [hunk 1]
```diff
@@ -564,6 +570,10 @@ namespace {
                 state.saturation[pu.phase_pos[ Oil ]] = so;
             }
         }
+
+        if( nextvar == int(vars.size()) )
+            return state;
+
         // Qs.
```
Comment (atgeirr) [inline opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp]: I think this is a bit confusing. It is correct, but I'd much rather have:
```
if (wellsActive()) {
    // Assign state.qs and state.bhp as before
    ...
} else {
    state.qs = ADB::null();
    state.bhp = ADB::null();
}
```
That would leave the function with a single return statement, which is preferable, also it is better to have the explicit `if (wellsActive())` than the obscure `if (nextvar == vars.size())`.

**Comment by alfbr** [comment]
Actually, being able to run without wells is quite useful. It enables simple initialisation of models which is sufficient for STOOIP estimation. Moreover, I hope to be able to contribute a number of integration tests this year, and quite a few will not need wells.

File: opm/autodiff/SimulatorFullyImplicitBlackoil_impl.hpp [hunk 2]
```diff
@@ -483,13 +483,15 @@ namespace Opm
...
         {
             std::vector<int> resv_prod;
-
-            for (int w = 0, nw = wells.number_of_wells; w < nw; ++w) {
-                if (is_resv_prod(wells, w) ||
-                    ((wells.name[w] != 0) &&
-                     is_resv_prod(wmap, wells.name[w], step)))
-                {
-                    resv_prod.push_back(w);
+            if( &wells )
```
Comment (atgeirr) [inline opm/autodiff/SimulatorFullyImplicitBlackoil_impl.hpp]: This must be wrong I think. The pointer to a reference must be nonzero, so this is a noop test. I think you must either pass a pointer to this function or place its call inside an `if` statement.

**Comment by atgeirr** [comment]
Apart from my two points above, I think this can be included, and I will merge it when it is updated. Thanks for the work Robert! I agree that a well-refactoring may be in our future, but not now.

One annoyance: there were whitespace changes mixed with the functional changes here, which makes it harder to review. Viewing with the semi-secret `?w=1` url helps, but not completely since it seems one cannot make comments while in that mode.

File: opm/autodiff/SimulatorFullyImplicitBlackoil_impl.hpp (hunk 2 above)
Comment (dr-robertk) [inline opm/autodiff/SimulatorFullyImplicitBlackoil_impl.hpp]: Well, every day you learn something ;-) and that's the new thing for today. 
You can easily do 

const Wells& wells = *((Wells *) 0);

That is basically what happens when the *wells is passed to the Simulator without checking whether it's NULL or not.

Comment (atgeirr) [inline opm/autodiff/SimulatorFullyImplicitBlackoil_impl.hpp]: Will not that lead to a segfault? Perhaps only when you use the reference?

Anyway: I think that the compiler would be allowed to assume that the pointer is nonzero, or perhaps it is undefined behaviour. It makes me very nervous, because it breaks a very fundamental rule: do not dereference a null pointer. Are you sure this is guaranteed to work by the standard? Even if this is correct, it is so surprising that we should not do it like this.

**Comment by dr-robertk** [comment]
I addressed the mentioned changes above and rebased. I checked SPE9 and a test called EQUILIBRIUM in my opm-data fork. Both seem to work.

File: opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp [hunk 3]
```diff
@@ -490,21 +494,30 @@ namespace {
...
+            vars0.push_back(qs);
+
+            // Initial well bottom-hole pressure.
+            assert (not xw.bhp().empty());
+            const V bhp = Eigen::Map<const V>(& xw.bhp()[0], xw.bhp().size());
+            vars0.push_back(bhp);
+        }
+        else
+        {
+            const V zero = ADB::null().value();
```
Comment (atgeirr) [inline opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp]: This is just V() I think, but ok...

File: opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp [hunk 4]
```diff
@@ -564,14 +577,13 @@ namespace {
...
             }
         }
+
         // Qs.
         state.qs = vars[ nextvar++ ];
 
         // Bhp.
         state.bhp = vars[ nextvar++ ];
 
-        assert(nextvar == int(vars.size()));
```
Comment (atgeirr) [inline opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp]: Your solution is now fine, but why remove the assert?

File: opm/autodiff/SimulatorFullyImplicitBlackoil_impl.hpp (hunk 2 above)
Comment (bska) [inline opm/autodiff/SimulatorFullyImplicitBlackoil_impl.hpp]: >>
```C++
const Wells& wells = *((Wells *) 0);
```
>
> Will not that lead to a segfault?

Well, the above is undefined behaviour so anything can happen.  Segmentation faults are the most common outcome.

File: opm/autodiff/FullyImplicitBlackoilSolver.hpp [hunk 5]
```diff
@@ -194,6 +194,12 @@ namespace Opm {
         std::vector<int>         primalVariable_;
 
         // Private methods.
+
+        // return true if wells are available
+        bool wellsActive() const { return wells_ ? wells_->number_of_wells > 0 : false ; }
+        // return wells object
+        const Wells& wells () const { assert( wells_ ); return *wells_; }
```
Comment (bska) [inline opm/autodiff/FullyImplicitBlackoilSolver.hpp]: The statement
```C++
assert (wells_);
```
is, strictly speaking, undefined.  The input to `assert` must be an expression of scalar type with a boolean result, so you really need to write
```C++
assert (wells_ != 0);
```
instead.

**Comment by dr-robertk** [comment]
Ok, done. There was an assertion that I remove accidentally.

File: opm/autodiff/FullyImplicitBlackoilSolver.hpp (hunk 5 above)
Comment (dr-robertk) [inline opm/autodiff/FullyImplicitBlackoilSolver.hpp]: Could you make an example?

Comment (bska) [inline opm/autodiff/FullyImplicitBlackoilSolver.hpp]: > Could you make an example?

I don't think I understand what you mean.  The input to macro `assert` **must** be a boolean and the pointer expression `wells_` is not a boolean.  Said expression is **convertible** to boolean, but `assert` does not enforce a boolean context although most typical implementations do.

So, either write `assert (wells_ != 0)` or circumvent the entire issue by writing
```C++
assert (wellsActive())
```

Comment (atgeirr) [inline opm/autodiff/FullyImplicitBlackoilSolver.hpp]: Are you sure? The webpage http://www.cplusplus.com/reference/cassert/assert/ gives
```
void assert (int expression);
```
as the form of the macro, but I guess it may be specified differently in the standard? If this really is illegal, I fear I have done a few assert(ptr) errors over the years... Of course there is no harm in using your suggested alternatives anyway.

**Comment by dr-robertk** [comment]
I addressed the assert( wells_ ) issue.

File: opm/autodiff/FullyImplicitBlackoilSolver.hpp (hunk 5 above)
Comment (dr-robertk) [inline opm/autodiff/FullyImplicitBlackoilSolver.hpp]: I meant an example where assert( wells_ ) would not work. I haven't come across any in the last 12 years I haven been working with C++ (different version on different platforms).

**Comment by atgeirr** [comment]
OK, merging.

File: opm/autodiff/FullyImplicitBlackoilSolver.hpp (hunk 5 above)
Comment (bska) [inline opm/autodiff/FullyImplicitBlackoilSolver.hpp]: > Are you sure?

Yes

>
```C++
void assert (int expression);
```

That's wrong.  `Assert` is a macro and therefore doesn't have a formal prototype/declaration.  The `int` part really means that you must pass an `int` (or, in C++, a `bool`).  There is no automatic type conversion going on.

> If this really is illegal, I fear I have done a few `assert(ptr)` errors over the years

It's a common mistake.  I've done it myself on numerous occasions.  Furthermore, the situation is even worse in C than in C++.  Of course, the typical outcome is that the statement behaves exactly in way the programmer expects...

> I meant an example where `assert( wells_ )` would not work.

Not off the top of my head, no.  I suppose a machine with a segmented architecture such as the (defunct) Data General Eclipse would be a candidate.  That machine had different register sets for pointers and integers so you would get more or less a random integer in the `assert` expression unless you explicitly converted the value to an integer.

**Comment by bska** [comment]
> I addressed the `assert( wells_ )` issue.

Thanks.  To be fair you didn't need the explicit `bool` conversion because `wells_ != 0` is already `bool`, but it doesn't hurt either.

File: opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp [hunk 6]
```diff
@@ -1254,6 +1275,17 @@ namespace {
         struct Chop01 {
             double operator()(double x) const { return std::max(std::min(x, 1.0), 0.0); }
         };
+
+        double infinityNorm( const ADB& a )
+        {
+            if( a.value().size() > 0 ) {
+                return a.value().matrix().template lpNorm<Eigen::Infinity> ();
```
Comment (bska) [inline opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp]: This breaks the build on GCC 4.4 and GCC 4.6 (although not on GCC 4.8).  You can't use the `.template` construction outside templates.  Specifically, neither `a` nor its `value().matrix()` depend on a template parameter so the use of `.template` is neither required nor permitted.

Of course, this was wrong in the initial version too, the one you refactored, but I didn't pick up on it until now.

Comment (atgeirr) [inline opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp]: Actually it was correct in the initial version since there it was in a template context. I did not notice this either, I guess that's what we have compilers for... I am surprised that 4.8 does not complain, my clang accepts it also. Does that indicate that this has been relaxed a little in C++11 or C++14? It does not matter much since we must compile on gcc 4.4 anyway.

Comment (bska) [inline opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp]: > Actually it was correct in the initial version since there it was in a template context

That's a necessary condition, not sufficient.  The `.template` construct is only appropriate if the expression of which it is part depends on a template parameter.  That was not the case in the original version either.

> I am surprised that 4.8 does not complain, my clang accepts it also. Does that indicate that this has been relaxed a little in C++11 or C++14?

Possibly.  I haven't read up on all the details.

Comment (notspiff) [inline opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp]: exactly that, see http://en.cppreference.com/w/cpp/language/dependent_name near the end.
